File: 38.外观数列.go
/*
 * @lc app=leetcode.cn id=38 lang=golang
 *
 * [38] 外观数列
 */

// @lc code=start
var ss []string
func countAndSay(n int) string {
	if n == 1 {
		ss = append(ss, string("1"))
	}
	if n <= len(ss) {
		return ss[n-1]
	}
	for len(ss) < n {
		ss = append(ss, describe(ss[len(ss)-1]))
	}
	return ss[n-1]
}

func describe(s string) string {
	var result string
	i := 0
	for j := i; i < len(s); {
		for j < len(s) && s[j] == s[i] {
			j++
		}
		n := strconv.Itoa(j-i)
		result += n
		result += string(s[i])
		i = j
	}
	return result
}
// @lc code=end

File: 38.外观数列.js
/*
 * @lc app=leetcode.cn id=38 lang=javascript
 *
 * [38] 外观数列
 */

// @lc code=start
/**
 * @param {number} n
 * @return {string}
 */
let ss = [];
var countAndSay = function(n) {
    if (n === 1) {
        ss.push('1');
    }
    if (ss.length === n) {
        return ss[n - 1];
    } 
    for (;ss.length < n;) {
        ss.push(describe(ss[ss.length - 1]));
    }
    return ss[n - 1]
};

let describe = function(s) {
    let result = '';
    let i = 0;
    for (let j = i; i < s.length; ) {
        for (; j < s.length && s[j] === s[i]; ) {
            j++;
        }
        result += (j-i).toString();
        result += s[i];
        i = j;
    } 
    return result;
};
// @lc code=end

